nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -357,6 +359,75 @@ impl<'a> Planner<'a> {
+        // version numbers consist of the first *two* components, which
+        // represent the year and the Nth release that year. So the major
+        // version in "22.2.7" is "22.2".
+        //
+        // A given major version of CockroachDB is backward compatible with the
+        // storage format of the previous major version of CockroachDB. This is
+        // shown by the `version` setting, which displays the current storage
+        // format version. When `version` is '22.2', versions v22.2.x or v23.1.x
+        // can be used to run a node. This allows for rolling upgrades of nodes
+        // within the cluster and also preserves the ability to rollback until
+        // the new software version can be validated.
+        //
+        // By default, when all nodes of a cluster are upgraded to a new major
+        // version, the upgrade is "auto-finalized"; `version` is changed to the
+        // new major version, and rolling back to a previous major version of
+        // CockroachDB is no longer possible.
+        //
+        // The `cluster.preserve_downgrade_option` setting can be used to
+        // control this. This setting can only be set to the current value
+        // of the `version` setting, and when it is set, CockroachDB will not
+        // perform auto-finalization. To perform finalization and finish the
+        // upgrade, a client must reset the "preserve downgrade option" setting.
+        // Finalization occurs in the background, and the "preserve downgrade
+        // option" setting should not be changed again until finalization
+        // completes.
+        //
+        // We determine the appropriate value for `preserve_downgrade_option`
+        // based on:
+        //
+        // 1. the _target_ cluster version from the `Policy` (what we want to
+        //    be running)
+        // 2. the `version` setting reported by CockroachDB (what we're
+        //    currently running)
+        //
+        // by saying:
+        //
+        // - If our target version matches what CockroachDB reports, we will
+        //   ensure `preserve_downgrade_option` is set to the same value. This
+        //   prevents auto-finalization when we deploy the next major version of
+        //   CockroachDB as part of an update.
+        // - If our target version does not match what CockroachDB reports,
+        //   we will ensure `preserve_downgrade_option` is set to the default
+        //   value (the empty string). This will trigger finalization.
+
+        let policy = self.input.target_cockroachdb_cluster_version();
+        let CockroachDbSettings { version, .. } =
+            self.input.cockroachdb_settings();
+
+        let value = if policy == version {
+            // Ensure `cluster.preserve_downgrade_option` is set
+            policy
+        } else {
+            // Ensure `cluster.preserve_downgrade_option` is reset
+            ""
```

Comment:
I think the answer to this is "no" but wanted to ask just in case - should we check here for "`policy` is older than `version`"? If I understand correctly, that should be impossible (that means we think we want to run a CRDB major version that is behind the current cluster version), so I'm guessing it's not worth checking for, since such a situation would blow up long before we got here.

Reply:
Based on the conversations in #oxide-update it sounds like we want to use the policy mechanism to hold back cluster versions as we upgrade CockroachDB, but to use the newer cluster version for any newly-initialized racks. In that case, "`policy` is older than `version`" is something that will indeed happen. As part of converting the value of `cockroachdb_settings_preserve_downgrade` to an enum, I was able to make this distinction a lot more clear and handle this case.